include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -433,6 +437,48 @@ auto applyDeflation(const SizeType i_begin, const SizeType i_end, RhoSender&& rh
          ex::ensure_started();
 }
 
+void initSortByColTypeIndex(SizeType i_begin, SizeType i_end,
+                            Matrix<const ColType, Device::CPU>& coltypes,
+                            Matrix<SizeType, Device::CPU>& index) {
+  namespace ex = pika::execution::experimental;
+  namespace di = dlaf::internal;
+
+  const SizeType n = problemSize(i_begin, i_end, index.distribution());
+  TileCollector tc{i_begin, i_end};
+
+  ex::start_detached(
+      ex::when_all(ex::when_all_vector(tc.read(coltypes)), ex::when_all_vector(tc.readwrite(index))) |
+      di::transform(di::Policy<Backend::MC>{}, [n](auto&& coltype_tiles, auto&& index_tiles) {
+        auto coltype_index = [](const ColType coltype) -> std::size_t {
+          switch (coltype) {
+            case ColType::UpperHalf:
+              return 0;
+            case ColType::Dense:
+              return 1;
+            case ColType::LowerHalf:
+              return 2;
+            case ColType::Deflated:
+              return 3;
+          }
+          return DLAF_UNREACHABLE(std::size_t);
+        };
+
+        const ColType* types = coltype_tiles[0].get().ptr();
+        std::array<std::size_t, 4> offsets{0, 0, 0, 0};
+        std::for_each(types, types + n, [&offsets, &coltype_index](const auto& coltype) {
+          if (coltype != ColType::Deflated)
+            offsets[1 + coltype_index(coltype)]++;
+        });
```

Comment:
If I'm not mistaken this is a reduction. `std::reduce` doesn't help here because it requires the elements that are being reduced and the type into which we are reducing to be the same. You could rewrite this with `std::accumulate`, but I'll leave it up to you to judge if it's more readable: https://godbolt.org/z/Pxj6hq81v (alternative: https://godbolt.org/z/vKd66zc9c). It's not great that `offsets` has to be wrapped in `ref` so I'm not sure it's a win. The main win is that it says `accumulate` which tells me right away that it's a reduction.

Reply:
Yes, thanks for the suggestion!

Actually I had already quickly checked if I had any `std`-way of doing this, but I was not able to come up with something really neat.

Even if I agree that with `std::accumulate` the name is a good hint for understanding what's going on, to me it seems that in this case it is not of much help because the init value is an array of initvalues, and the binary operator has to do everything exactly as in the current solution: select the right counter and increment it (+ boilerplate to make accumulate happy, e.g. return `std::ref` to the counters).

In the end, we would use `std::accumulate` just as a `std::for_each` with some more boilerplate...don't you think?

In any case, thank you for the different view on the problem!

Reply:
Yeah, agreed. I'm fine with leaving it as is.